library/core/src/ptr/mod.rs

```diff
@@ -65,6 +65,14 @@
 //! separate allocated object), heap allocations (each allocation created by the global allocator is
 //! a separate allocated object), and `static` variables.
 //!
+//! It is guaranteed that an allocated object never spans more than `isize::MAX` bytes.
+//! An allocated object cannot contain [`null()`] and cannot contain the last (`usize::MAX`) byte
+//! of the address space. As a consequence, any address which is either in the allocated object or
+//! one byte past the last address in the allocated object can be computed without wrapping around
+//! the address space.
```

Comment:
Nice. :)

I see only one potential problem: if the object has size zero, "one byte past the last address in the object" is ill-defined as there is no address in the object. I never really like this "one past the end" terminology either. This is just "base_addr + size", and I think of it as the pointer "at the end".

Reply:
Migrating this conversation to [the other thread](https://github.com/rust-lang/rust/pull/116675#discussion_r1382380661) so we have it all in one place.